File: 2025-2/sesion-de-clase-16/SoftProgPersistencia/DAOImpl/RRHH/AreaDAOImpl.cs
using System.Data;
using System.Data.Common;
using PUCP.SoftProg.Modelo.RRHH;
using PUCP.SoftProg.Persistencia.DAO.RRHH;

namespace PUCP.SoftProg.Persistencia.DAOImpl.RRHH {
    public class AreaDAOImpl : BaseDAOImpl<Area>, IAreaDAO {
        protected override DbCommand ComandoCrear(DbConnection conn, Area area) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "insertarArea";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_nombre", DbType.String, area.Nombre);
            this.AgregarParametroEntrada(cmd, "@p_activo", DbType.Boolean, area.IsActive);
            this.AgregarParametroSalida(cmd, "@p_id", DbType.Int32);

            return cmd;
        }

        protected override DbCommand ComandoActualizar(DbConnection conn, Area area) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "modificarArea";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_nombre", DbType.String, area.Nombre);
            this.AgregarParametroEntrada(cmd, "@p_activo", DbType.Boolean, area.IsActive);
            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, area.Id);

            return cmd;
        }

        protected override DbCommand ComandoEliminar(DbConnection conn, int id) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "eliminarArea";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand ComandoLeer(DbConnection conn, int id) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "buscarAreaPorId";
            cmd.CommandType = CommandType.StoredProcedure;
            
            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand ComandoLeerTodos(DbConnection conn) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "listarAreas";
            cmd.CommandType = CommandType.StoredProcedure;
            return cmd;
        }

        protected override Area MapearModelo(DbDataReader reader) {
            return new Area {
                Id = (int)reader["id"], 
                Nombre = reader["nombre"].ToString(),
                IsActive = (bool)reader["activo"],
            };
        }
    }
}

File: 2025-2/sesion-de-clase-19/java/SoftProg/SoftProgPersistencia/src/main/java/pe/edu/pucp/inf30/softprog/daoimpl/rrhh/AreaDAOImpl.java
package pe.edu.pucp.inf30.softprog.daoimpl.rrhh;

import java.sql.CallableStatement;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Types;
import pe.edu.pucp.inf30.softprog.dao.rrhh.AreaDAO;
import pe.edu.pucp.inf30.softprog.daoimpl.BaseDAO;
import pe.edu.pucp.inf30.softprog.modelo.rrhh.Area;

/**
 *
 * @author eric
 */
public class AreaDAOImpl extends BaseDAO<Area> implements AreaDAO {    
    @Override
    protected PreparedStatement comandoCrear(Connection conn, 
            Area modelo) throws SQLException {
        
        String sql = "{call insertarArea(?, ?, ?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setString("p_nombre", modelo.getNombre());
        cmd.setBoolean("p_activo", modelo.isActivo());
        cmd.registerOutParameter("p_id", Types.INTEGER);
        
        return cmd;
    }

    @Override
    protected PreparedStatement comandoActualizar(Connection conn, 
            Area modelo) throws SQLException {
        
        String sql = "{call modificarArea(?, ?, ?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setString("p_nombre", modelo.getNombre());
        cmd.setBoolean("p_activo", modelo.isActivo());
        cmd.setInt("p_id", modelo.getId());
        
        return cmd;
    }

    @Override
    protected PreparedStatement comandoEliminar(Connection conn, 
            Integer id) throws SQLException {
        
        String sql = "{call eliminarArea(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt("p_id", id);
        
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeer(Connection conn, 
            Integer id) throws SQLException {
        
        String sql = "{call buscarAreaPorId(?)}";
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt("p_id", id);
        
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeerTodos(
            Connection conn) throws SQLException {
        
        String sql = "{call listarAreas()}";
        CallableStatement cmd = conn.prepareCall(sql);
        
        return cmd;
    }

    @Override
    protected Area mapearModelo(ResultSet rs) throws SQLException {
        Area modelo = new Area();
        
        modelo.setId(rs.getInt("id"));
        modelo.setNombre(rs.getString("nombre"));
        modelo.setActivo(rs.getBoolean("activo"));
        
        return modelo;
    }
}
